File: mobile/android/TrainBlink/app/src/main/java/org/trainblink/app/network/ApiService.kt
package org.trainblink.app.network

import android.content.Context
import com.google.gson.GsonBuilder
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import org.trainblink.app.BuildConfig
import org.trainblink.app.models.*
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * API Service interface for backend communication
 */
interface ApiService {

    // MARK: - Trip Endpoints

    @POST("/api/v1/trips/start")
    suspend fun startTrip(@Body request: StartTripRequest): TripResponse

    @POST("/api/v1/trips/end")
    suspend fun endTrip(@Body request: EndTripRequest): TripResponse

    @GET("/api/v1/trips/active")
    suspend fun getActiveTrip(): TripResponse

    @GET("/api/v1/trips")
    suspend fun getUserTrips(
        @Query("limit") limit: Int = 10,
        @Query("offset") offset: Int = 0
    ): TripsListResponse

    @DELETE("/api/v1/trips/{id}")
    suspend fun cancelTrip(@Path("id") tripId: UUID): TripResponse

    @PATCH("/api/v1/trips/{id}/discovery")
    suspend fun updateDiscoveryEnabled(
        @Path("id") tripId: UUID,
        @Body request: UpdateDiscoveryRequest
    ): TripResponse

    // MARK: - Discovery Endpoints

    @POST("/api/v1/discoveries/log")
    suspend fun logDiscovery(@Body request: LogDiscoveryRequest)

    @GET("/api/v1/discoveries/stats")
    suspend fun getDiscoveryStats(): DiscoveryStatsResponse

    // MARK: - Matrix Endpoints

    @POST("/api/v1/matrix/dm/create")
    suspend fun createEphemeralDM(@Body request: CreateEphemeralDMRequest): MatrixRoomResponse

    @GET("/api/v1/matrix/dm/active")
    suspend fun getActiveEphemeralDMs(): MatrixRoomsResponse

    @PATCH("/api/v1/matrix/dm/{id}/extend")
    suspend fun extendRoomLifetime(
        @Path("id") roomId: UUID,
        @Body request: ExtendRoomRequest
    ): Map<String, Any>

    companion object {
        /**
         * Create API service instance
         */
        fun create(context: Context, userId: UUID? = null, jwtToken: String? = null): ApiService {
            val gson = GsonBuilder()
                .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
                .create()

            // Create authentication interceptor
            val authInterceptor = Interceptor { chain ->
                val original = chain.request()
                val requestBuilder = original.newBuilder()

                // Add user ID header (development)
                userId?.let {
                    requestBuilder.header("X-User-ID", it.toString())
                }

                // Add JWT token (production)
                jwtToken?.let {
                    requestBuilder.header("Authorization", "Bearer $it")
                }

                val request = requestBuilder.build()
                chain.proceed(request)
            }

            // Create logging interceptor
            val loggingInterceptor = HttpLoggingInterceptor().apply {
                level = if (BuildConfig.DEBUG) {
                    HttpLoggingInterceptor.Level.BODY
                } else {
                    HttpLoggingInterceptor.Level.NONE
                }
            }

            // Create OkHttp client
            val okHttpClient = OkHttpClient.Builder()
                .addInterceptor(authInterceptor)
                .addInterceptor(loggingInterceptor)
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build()

            // Create Retrofit instance
            val retrofit = Retrofit.Builder()
                .baseUrl(BuildConfig.API_BASE_URL)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()

            return retrofit.create(ApiService::class.java)
        }
    }
}

// MARK: - Matrix Request/Response Models

/**
 * Request to create ephemeral DM
 */
data class CreateEphemeralDMRequest(
    @SerializedName("discovered_user_ble_id")
    val discoveredUserBLEID: String,
    @SerializedName("mls_group_id")
    val mlsGroupId: String? = null
)

/**
 * Request to extend room lifetime
 */
data class ExtendRoomRequest(
    @SerializedName("extension_hours")
    val extensionHours: Int
)

/**
 * Matrix ephemeral room model
 */
data class MatrixEphemeralRoom(
    val id: UUID,
    @SerializedName("room_id")
    val roomId: String,
    @SerializedName("trip1_id")
    val trip1Id: UUID,
    @SerializedName("trip2_id")
    val trip2Id: UUID,
    @SerializedName("anonymous_id_1")
    val anonymousId1: String,
    @SerializedName("anonymous_id_2")
    val anonymousId2: String,
    @SerializedName("mls_group_id")
    val mlsGroupId: String?,
    @SerializedName("expires_at")
    val expiresAt: java.util.Date,
    @SerializedName("message_count")
    val messageCount: Int,
    @SerializedName("last_message_at")
    val lastMessageAt: java.util.Date?,
    @SerializedName("created_at")
    val createdAt: java.util.Date
) {
    /**
     * Time remaining until expiry in milliseconds
     */
    val timeRemaining: Long
        get() = maxOf(0, expiresAt.time - java.util.Date().time)

    /**
     * Check if room is expiring soon (within 1 hour)
     */
    val isExpiringSoon: Boolean
        get() = timeRemaining < 3600_000

    /**
     * Format time remaining
     */
    val timeRemainingFormatted: String
        get() {
            val hours = (timeRemaining / (1000 * 60 * 60)).toInt()
            val minutes = ((timeRemaining / (1000 * 60)) % 60).toInt()

            return when {
                hours > 0 -> "${hours}h ${minutes}m"
                minutes > 0 -> "${minutes}m"
                else -> "Expiring soon"
            }
        }
}

/**
 * API response for single Matrix room
 */
data class MatrixRoomResponse(
    val status: String,
    val data: MatrixEphemeralRoom?
)

/**
 * API response for Matrix room list
 */
data class MatrixRoomsResponse(
    val status: String,
    val data: RoomsData?
) {
    data class RoomsData(
        val rooms: List<MatrixEphemeralRoom>,
        val total: Int
    )
}

File: mobile/ios/TrainBlink/TrainBlink/Services/APIService.swift
//
//  APIService.swift
//  TrainBlink
//
//  HTTP API client for TrainBlink backend
//

import Foundation
import Alamofire

/// API Service for backend communication
class APIService {

    // MARK: - Configuration

    private let baseURL: String
    private let session: Session
    private var userId: UUID?
    private var jwtToken: String?

    // MARK: - Initialization

    init(baseURL: String = "http://localhost:8080") {
        self.baseURL = baseURL
        self.session = Session()
    }

    /// Set user ID for authenticated requests
    func setUserId(_ userId: UUID) {
        self.userId = userId
    }

    /// Set JWT token for production authentication
    func setJWTToken(_ token: String) {
        self.jwtToken = token
    }

    // MARK: - Headers

    private var headers: HTTPHeaders {
        var headers = HTTPHeaders()

        // Development: Use X-User-ID header
        if let userId = userId {
            headers.add(name: "X-User-ID", value: userId.uuidString)
        }

        // Production: Use JWT token
        if let token = jwtToken {
            headers.add(name: "Authorization", value: "Bearer \(token)")
        }

        headers.add(name: "Content-Type", value: "application/json")
        return headers
    }

    // MARK: - Trip Endpoints

    /// Start a new trip
    func startTrip(request: StartTripRequest) async throws -> Trip {
        let response = try await session
            .request("\(baseURL)/api/v1/trips/start", method: .post, parameters: request, encoder: JSONParameterEncoder.default, headers: headers)
            .validate()
            .serializingDecodable(TripResponse.self)
            .value

        guard let trip = response.data else {
            throw APIError.invalidResponse
        }

        return trip
    }

    /// End current trip
    func endTrip(tripId: UUID) async throws {
        let request = EndTripRequest(tripId: tripId)

        try await session
            .request("\(baseURL)/api/v1/trips/end", method: .post, parameters: request, encoder: JSONParameterEncoder.default, headers: headers)
            .validate()
            .serializingDecodable(TripResponse.self)
            .value
    }

    /// Get active trip
    func getActiveTrip() async throws -> Trip? {
        let response = try await session
            .request("\(baseURL)/api/v1/trips/active", method: .get, headers: headers)
            .validate()
            .serializingDecodable(TripResponse.self)
            .value

        return response.data
    }

    /// Get user trips
    func getUserTrips(limit: Int = 10, offset: Int = 0) async throws -> [Trip] {
        let response = try await session
            .request("\(baseURL)/api/v1/trips", method: .get, parameters: ["limit": limit, "offset": offset], headers: headers)
            .validate()
            .serializingDecodable(TripsListResponse.self)
            .value

        return response.data?.trips ?? []
    }

    /// Cancel trip
    func cancelTrip(tripId: UUID) async throws {
        try await session
            .request("\(baseURL)/api/v1/trips/\(tripId.uuidString)", method: .delete, headers: headers)
            .validate()
            .serializingDecodable(TripResponse.self)
            .value
    }

    /// Update discovery enabled
    func updateDiscoveryEnabled(tripId: UUID, enabled: Bool) async throws {
        try await session
            .request("\(baseURL)/api/v1/trips/\(tripId.uuidString)/discovery", method: .patch, parameters: ["discovery_enabled": enabled], encoder: JSONParameterEncoder.default, headers: headers)
            .validate()
            .serializingDecodable(TripResponse.self)
            .value
    }

    // MARK: - Discovery Endpoints

    /// Log discovery event
    func logDiscovery(request: LogDiscoveryRequest) async throws {
        try await session
            .request("\(baseURL)/api/v1/discoveries/log", method: .post, parameters: request, encoder: JSONParameterEncoder.default, headers: headers)
            .validate()
            .serializingData()
            .value
    }

    /// Get discovery stats
    func getDiscoveryStats() async throws -> DiscoveryStatsResponse.DiscoveryStats? {
        let response = try await session
            .request("\(baseURL)/api/v1/discoveries/stats", method: .get, headers: headers)
            .validate()
            .serializingDecodable(DiscoveryStatsResponse.self)
            .value

        return response.data
    }

    // MARK: - Matrix Endpoints

    /// Create ephemeral DM
    func createEphemeralDM(discoveredUserBLEID: String, mlsGroupID: String? = nil) async throws -> MatrixEphemeralRoom {
        let params: [String: Any] = [
            "discovered_user_ble_id": discoveredUserBLEID,
            "mls_group_id": mlsGroupID as Any
        ]

        let response = try await session
            .request("\(baseURL)/api/v1/matrix/dm/create", method: .post, parameters: params, encoding: JSONEncoding.default, headers: headers)
            .validate()
            .serializingDecodable(MatrixRoomResponse.self)
            .value

        guard let room = response.data else {
            throw APIError.invalidResponse
        }

        return room
    }

    /// Get active ephemeral DMs
    func getActiveEphemeralDMs() async throws -> [MatrixEphemeralRoom] {
        let response = try await session
            .request("\(baseURL)/api/v1/matrix/dm/active", method: .get, headers: headers)
            .validate()
            .serializingDecodable(MatrixRoomsResponse.self)
            .value

        return response.data?.rooms ?? []
    }

    /// Extend room lifetime
    func extendRoomLifetime(roomId: UUID, extensionHours: Int) async throws {
        try await session
            .request("\(baseURL)/api/v1/matrix/dm/\(roomId.uuidString)/extend", method: .patch, parameters: ["extension_hours": extensionHours], encoder: JSONParameterEncoder.default, headers: headers)
            .validate()
            .serializingData()
            .value
    }
}

// MARK: - Matrix Response Models

struct MatrixEphemeralRoom: Codable, Identifiable {
    let id: UUID
    let roomId: String
    let trip1Id: UUID
    let trip2Id: UUID
    let anonymousId1: String
    let anonymousId2: String
    let mlsGroupId: String?
    let expiresAt: Date
    let messageCount: Int
    let lastMessageAt: Date?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case roomId = "room_id"
        case trip1Id = "trip1_id"
        case trip2Id = "trip2_id"
        case anonymousId1 = "anonymous_id_1"
        case anonymousId2 = "anonymous_id_2"
        case mlsGroupId = "mls_group_id"
        case expiresAt = "expires_at"
        case messageCount = "message_count"
        case lastMessageAt = "last_message_at"
        case createdAt = "created_at"
    }

    /// Time remaining until expiry
    var timeRemaining: TimeInterval {
        return max(0, expiresAt.timeIntervalSince(Date()))
    }

    /// Check if room is expiring soon (within 1 hour)
    var isExpiringSoon: Bool {
        return timeRemaining < 3600
    }

    /// Format time remaining
    var timeRemainingFormatted: String {
        let hours = Int(timeRemaining) / 3600
        let minutes = Int(timeRemaining) % 3600 / 60

        if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else if minutes > 0 {
            return "\(minutes)m"
        } else {
            return "Expiring soon"
        }
    }
}

struct MatrixRoomResponse: Codable {
    let status: String
    let data: MatrixEphemeralRoom?
}

struct MatrixRoomsResponse: Codable {
    let status: String
    let data: RoomsData?

    struct RoomsData: Codable {
        let rooms: [MatrixEphemeralRoom]
        let total: Int
    }
}

// MARK: - Error Handling

enum APIError: Error, LocalizedError {
    case invalidResponse
    case networkError(String)
    case unauthorized
    case serverError(Int)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid response from server"
        case .networkError(let message):
            return "Network error: \(message)"
        case .unauthorized:
            return "Unauthorized request"
        case .serverError(let code):
            return "Server error: \(code)"
        }
    }
}
